cpp/src/groupby/new_groupby.cu

```diff
@@ -190,3 +195,220 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
+                          ctxt);
+    if (status != GDF_SUCCESS)
+      return status;
+
+    // lets filter out all the nulls in the group by key column by:
+    // we will take the data which has been sorted such that the nulls in the group by keys are all last
+    // then using the gdf_table's property of row-validity mask we can count how many rows have 
+    // a null in the group by keys and use that to resize the data
+    std::unique_ptr< gdf_table<int32_t> > group_by_keys_table{new gdf_table<int32_t>{num_groupby_cols, &orderby_cols_vect[0]}};
+    int valid_count;
+    status = group_by_keys_table->get_num_valid_rows(valid_count);
+    if (status != GDF_SUCCESS)
+      return status;
+
+    for (int i = 0; i < num_data_cols; i++)    {
+      data_cols_in[i]->size = valid_count;
+      data_cols_out[i]->size = valid_count;
+    }
+    
+    // run gather operation to establish new order
+    std::unique_ptr< gdf_table<int32_t> > table_in{new gdf_table<int32_t>{num_data_cols, data_cols_in}};
+    std::unique_ptr< gdf_table<int32_t> > table_out{new gdf_table<int32_t>{num_data_cols, data_cols_out}};
+    
+    status = table_in->gather<int32_t>(sorted_indices, *table_out.get());
+    if (status != GDF_SUCCESS)
+      return status;
+   
+    ctxt->flag_nulls_sort_behavior = flag_nulls_sort_behavior;
+
+    status = gdf_group_start_indices(num_data_cols, data_cols_out, num_groupby_cols,
+                      groupby_col_indices, group_start_indices, ctxt);
+
+    return status;
+  }
+}
+
+
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the group by columns,
+   * all input data is assumed to already be sorted by these group by columns. 
+   * This function calculates a list of the start indices of each group
+   *
+   * @Param[in] The number of columns in the dataset (assumed to already be sorted)
+   * @Param[in] The input columns in the dataset
+   * @Param[in] The number of columns to be grouping by
+   * @Param[in] The column indices of the input dataset that will be grouped by
+   * @Param[out] A column containing the starting indices of each group. Indices based off of new sort order. (needs to be pre-allocated)
+   * @Param[in] The context used to control how nulls are treated in a sort
+   *   context->flag_nulls_sort_behavior< 0 = Nulls are are treated as largest, 
+   *   1 = Nulls are treated as smallest, 2 = Special multi-sort case any row with null is largest>
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_start_indices(int num_data_cols,
+                     gdf_column** data_cols_in,
+									   int num_groupby_cols,
+									   int * groupby_col_indices,
+									   gdf_column* group_start_indices,
```

Comment:
This code assumes `groupby_col_indices->dtype == GDF_INT32`. Make sure to add a `GDF_REQUIRE` for that.